pkg/task/taskset.go

```diff
@@ -209,46 +206,81 @@ func (s *TaskSet) sortTaskGraph() *SortTaskGraphResult {
 				delete(missingTaskIdsInMap, task.UntypedID().ReferenceIDString())
 			}
 
-			// When there were no task runnable only with the missing sources,
-			// the task graph shape is at least cyclic.
-			hasCyclicDependencies := true
-			for taskId := range nonResolvedTasksMap {
-				canBeNextStartingPoint := true
-				for source := range currentMissingTaskDependencies[taskId] {
-					if _, exist := missingTaskIdsInMap[source]; !exist {
-						canBeNextStartingPoint = false
-						break
-					}
-				}
-				if canBeNextStartingPoint {
-					hasCyclicDependencies = false
-					break
-				}
-			}
-
-			if !hasCyclicDependencies {
-				for _, task := range nonResolvedTasksMap {
-					delete(missingTaskIdsInMap, task.UntypedID().ReferenceIDString())
-				}
-			}
-
 			missingSources := []taskid.UntypedTaskReference{}
 			for source := range missingTaskIdsInMap {
 				missingSources = append(missingSources, taskid.NewTaskReference[any](source))
 			}
 
+			if len(missingSources) == 0 {
+				for _, taskID := range sortedMapKeys(nonResolvedTasksMap) {
+					dependentFrom := map[string]string{} // A map tracks the path where the task depended from.
+					dependentFrom[taskID] = "START"
+					queue := map[string]struct{}{}
+					queue[taskID] = struct{}{}
+
+					for len(queue) > 0 {
+						nextTaskID := sortedMapKeys(queue)[0]
+						delete(queue, nextTaskID)
+						for dependency := range currentMissingTaskDependencies[nextTaskID] {
+							prevParent := ""
+							for visitedTask := range dependentFrom {
+								// The task ID contains implementation hash(#default), it should match with the prefix.
+								if strings.HasPrefix(visitedTask, dependency) {
+									prevParent = dependentFrom[visitedTask]
+									break
+								}
+							}
+							if prevParent != "" {
+								if prevParent == "START" {
+									// now we found the path to loop back to the START. trace back the cyclic path.
+									path := []string{}
+									queue := map[string]struct{}{}
+									queue[nextTaskID] = struct{}{}
+									for len(queue) > 0 {
+										nextTaskID := sortedMapKeys(queue)[0]
+										if nextTaskID == "START" {
+											break
+										}
+										delete(queue, nextTaskID)
+										path = append(path, nextTaskID)
+										queue[dependentFrom[nextTaskID]] = struct{}{}
+									}
+
+									return &SortTaskGraphResult{
+										Runnable:               false,
+										TopologicalSortedTasks: nil,
+										CyclicDependencyPath:   fmt.Sprintf("... -> %s] -> [%s] -> [%s -> ...", path[len(path)-1], strings.Join(path, " -> "), path[0]),
+										MissingDependencies:    missingSources,
+									}
+								}
+							} else {
+								for taskID := range nonResolvedTasksMap {
+									if strings.HasPrefix(taskID, dependency) {
+										dependentFrom[taskID] = nextTaskID
+										queue[taskID] = struct{}{}
+										break
+									}
+								}
+							}
+
+						}
+					}
+				}
+				panic("unreachable")
```

Comment:
I'm not sure here should panic without any logs.

It might be better to leave some hints if the logic reached here or returning error to handle this situation?

Reply:
Added comment and changed the panic message to contain more message.

Reply:
I'm sorry I meant it would be better to dump data for debugging if it happened?

So we can gather the debugging information from GitHub issues or from internal users/systems?

Reply:
I added some information on there. 
But I'm not sure if it's enough because I can't imagine the actual situation to reach there.